pkg/diff/diff.go

```diff
@@ -487,140 +528,187 @@ func (t *TableDiff) compareChecksum(ctx context.Context, chunk *ChunkRange) (boo
+			if err != nil {
+				return false, err
+			}
+		}
+
+		if lastTargetData == nil {
+			lastTargetData, err = getRowData(targetRows)
+			if err != nil {
+				return false, err
+			}
+		}
+
+		if lastSourceData == nil {
+			// don't have source data, so all the targetRows's data is redundant, should be deleted
+			for lastTargetData != nil {
+				sql := generateDML("delete", lastTargetData, t.TargetTable.info, t.TargetTable.Schema)
 				log.Info("[delete]", zap.String("sql", sql))
 				t.wg.Add(1)
 				t.sqlCh <- sql
 				equal = false
+
+				lastTargetData, err = getRowData(targetRows)
+				if err != nil {
+					return false, err
+				}
 			}
 			break
 		}
-		if index2 == len(rowsData2) {
-			// rowsData2 lack some data, should insert them
-			for ; index1 < len(rowsData1); index1++ {
-				sql := generateDML("replace", rowsData1[index1], t.TargetTable.info, t.TargetTable.Schema)
+
+		if lastTargetData == nil {
+			// target lack some data, should insert the last source datas
+			for lastSourceData != nil {
+				sql := generateDML("replace", lastSourceData, t.TargetTable.info, t.TargetTable.Schema)
 				log.Info("[insert]", zap.String("sql", sql))
 				t.wg.Add(1)
 				t.sqlCh <- sql
 				equal = false
+
+				lastSourceData, err = getSourceRow()
+				if err != nil {
+					return false, err
+				}
 			}
 			break
 		}
-		eq, cmp, err := compareData(rowsData1[index1], rowsData2[index2], orderKeyCols)
+
+		eq, cmp, err := compareData(lastSourceData, lastTargetData, orderKeyCols)
 		if err != nil {
 			return false, errors.Trace(err)
 		}
 		if eq {
-			index1++
-			index2++
+			lastSourceData = nil
+			lastTargetData = nil
 			continue
 		}
+
 		equal = false
 		switch cmp {
 		case 1:
 			// delete
-			sql := generateDML("delete", rowsData2[index2], t.TargetTable.info, t.TargetTable.Schema)
+			sql := generateDML("delete", lastTargetData, t.TargetTable.info, t.TargetTable.Schema)
 			log.Info("[delete]", zap.String("sql", sql))
 			t.wg.Add(1)
 			t.sqlCh <- sql
-			index2++
+			lastTargetData = nil
 		case -1:
 			// insert
-			sql := generateDML("replace", rowsData1[index1], t.TargetTable.info, t.TargetTable.Schema)
+			sql := generateDML("replace", lastSourceData, t.TargetTable.info, t.TargetTable.Schema)
 			log.Info("[insert]", zap.String("sql", sql))
 			t.wg.Add(1)
 			t.sqlCh <- sql
-			index1++
+			lastSourceData = nil
 		case 0:
 			// update
-			sql := generateDML("replace", rowsData1[index1], t.TargetTable.info, t.TargetTable.Schema)
+			sql := generateDML("replace", lastSourceData, t.TargetTable.info, t.TargetTable.Schema)
 			log.Info("[update]", zap.String("sql", sql))
 			t.wg.Add(1)
 			t.sqlCh <- sql
-			index1++
-			index2++
+			lastSourceData = nil
+			lastTargetData = nil
 		}
 	}
 
+	if equal {
+		log.Info("rows is equal", zap.String("table", dbutil.TableName(t.TargetTable.Schema, t.TargetTable.Table)), zap.String("where", dbutil.ReplacePlaceholder(chunk.Where, chunk.Args)), zap.Duration("cost", time.Since(beginTime)))
```

Comment:
Does this output too many log for a huge table?

Reply:
will only print for each chunk, I think this log is necessary, users can know what diff is doing now, and also know the time cost.